### PR TITLE
Added 'extra_install_args' module option to allow extra upgrade/install specific zypper arguments

### PR DESCRIPTION
**Pull request by marbud0**
Add `extra_install_args` to zypper to allow for installation / upgrade specific arguments.

Example install / upgrade zypper arguments that faisl using the `extra_args`

* --allow-vendor-change
* --replacefiles and
* --force-resolution

##### SUMMARY

Some zypper installation / upgrade options could not be used with the zypper module. Added a new `extra_install_args` argument to the module to handle install / upgrade specific zypper arguments.

Fixes issue #381

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

zypper

##### ADDITIONAL INFORMATION

Please issue #381

```paste below

```

### CODE REVIEW

**Comment by felixfontein** [comment]
I'm not sure whether this is a good idea. Such options generally make it impossible to do proper idempotence checking. Adding such options is generally not seen as good style for Ansible modules.

(Also, there's already a module option for `--force-resolution`. And if you add such an option, make it a list of strings, and not a string.)

**Comment by marbud0** [comment]
> I'm not sure whether this is a good idea. Such options generally make it impossible to do proper idempotence checking. Adding such options is generally not seen as good style for Ansible modules.

I can understand your point. Still without some way of using the whole spectrum of zypper options via the zypper module we would be forced to use zypper commands directly.

> (And if you add such an option, make it a list of strings, and not a string.)
I don't follow you here.. Could you elaborate?

**Comment by dcermak** [comment]
> > (And if you add such an option, make it a list of strings, and not a string.)
> 
> I don't follow you here.. Could you elaborate?

I think what @felixfontein meant is that you can write this:
```yaml
- zypper:
    name: nmap
    state: present
    extra_install_args:
      - allow-vendor-change
      - force-resolution
```
and not:
```yaml
- zypper:
    name: nmap
    state: present
    extra_install_args: "--allow-vendor-change --force-resolution"
```

**Comment by felixfontein** [comment]
Yes, that's what I mean (though with `- "--allow-vendor-change"` in the list entries, not `- allow-vendor-change`, i.e. as it is implemented now in this PR). I would still not add such an option though. It's better to support individual arguments with their own options, especially since I'd assume that zypper isn't getting a bunch of new options every release.

File: plugins/modules/packaging/os/zypper.py [hunk 1]
```diff
@@ -106,6 +106,12 @@
         description:
           - Add additional options to C(zypper) command.
           - Options should be supplied in a single line as if given in the command line.
+    extra_install_args:
+        version_added: "2.9"
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper.py:110]:
```suggestion
        version_added: "2.10"
        type: list
        elements: str
```

File: plugins/modules/packaging/os/zypper.py [hunk 2]
```diff
@@ -478,6 +495,7 @@ def main():
             update_cache=dict(required=False, aliases=['refresh'], default='no', type='bool'),
             oldpackage=dict(required=False, default='no', type='bool'),
             extra_args=dict(required=False, default=None),
+            extra_install_args=dict(type='list', required=False, default=None),
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper.py:498]:
```suggestion
            extra_install_args=dict(type='list', elements='str', required=False, default=None),
```

File: plugins/modules/packaging/os/zypper.py [hunk 3]
```diff
@@ -188,7 +209,7 @@
 import xml
 import re
 from xml.dom.minidom import parseString as parseXML
-from ansible.module_utils.six import iteritems
+# from ansible.module_utils.six import iteritems  # Remove as it is not used
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper.py:212]: Could you simply delete this line?

**Comment by felixfontein** [comment]
(You can ignore the ansible/third-party-check failed test. It's unrelated to this PR.)

**Comment by marbud0** [comment]
> (You can ignore the ansible/third-party-check failed test. It's unrelated to this PR.)

OK.. Thanks.. I was scratching my head over that..

**Comment by felixfontein** [comment]
You might also want to use these options somewhere in the tests, so we know that they don't break anything. (Having extra tests for them would be even nicer, though I guess it's too much work.)

File: tests/integration/targets/zypper/tasks/zypper.yml [hunk 4]
```diff
@@ -442,3 +442,23 @@
...
+- name: Install gnu-netcat with force_resolution and replacefiles
+  zypper:
+    name: gnu-netcat
+    state: present
+    force_resolution: true
+    replacefiles: true
+
+- name: Dist upgrade with allow_vendor_change
+  zypper:
+    name: '*'
```
Comment (felixfontein) [inline tests/integration/targets/zypper/tasks/zypper.yml:462]: Please limit this to something: a) this test now downlodas *a lot* of packages, and b) it upgrades the Python 2 installation, thus modifying the environment.

Comment (marbud0) [reply tests/integration/targets/zypper/tasks/zypper.yml]: The dist-upgrade state mandate '*' as name.. So I just removed it for now..

Comment (felixfontein) [reply tests/integration/targets/zypper/tasks/zypper.yml]: Ah, too bad. In that case, not testing it for now is probably best...

**Comment by felixfontein** [comment]
@dcermak do you also want to take a look, or is it ok for you if we merge?
@evrardjp are the tests ok for you?

**Comment by evrardjp** [comment]
I can ensure it's a good test a little later today, but at first sight it's good.

File: tests/integration/targets/zypper/tasks/zypper.yml [hunk 5]
```diff
@@ -442,3 +442,17 @@
...
+      - -gnu-netcat
+      - +netcat-openbsd
+    state: present
+
+- name: Install gnu-netcat with force_resolution and replacefiles
+  zypper:
+    name: gnu-netcat
+    state: present
+    force_resolution: true
+    replacefiles: true
```
Comment (evrardjp) [inline tests/integration/targets/zypper/tasks/zypper.yml:458]: This is not a good test of replacefiles. The force_resolution has been tested above, so I am not sure those two tasks bring anything. Ideally we should craft a fixture with two non-conflicting rpms with conflicting files, copy and install them with replace files, and check the results.

Comment (marbud0) [reply tests/integration/targets/zypper/tasks/zypper.yml]: 59c7449 includes the above. But the tests "install netcat-openbsd which conflicts with gnu-netcat" fails. 
Are test environments re-used? Looks like a previous run left some packages behind causing an unexpected failure in the test.

File: plugins/modules/packaging/os/zypper.py [hunk 6]
```diff
@@ -164,7 +178,14 @@
   zypper:
     name: '*'
     state: dist-upgrade
-    extra_args: '--no-allow-vendor-change --allow-arch-change'
+    allow_vendor_change: true
+    extra_args: '--allow-arch-change'
+
+- name: Perform a installaion of nmap with the install option replacefiles and allow_vendor_change
```
Comment (evrardjp) [inline plugins/modules/packaging/os/zypper.py:184]: nit: you should remove allow_vendor_change in the name of the task.

File: tests/integration/targets/zypper/tasks/zypper.yml (hunk 5 above)
Comment (evrardjp) [reply tests/integration/targets/zypper/tasks/zypper.yml]: Maybe we should just add a removal of those packages first, to make sure we are clean.

File: plugins/modules/packaging/os/zypper.py (hunk 6 above)
Comment (marbud0) [reply plugins/modules/packaging/os/zypper.py]: Done in 78326ce

File: tests/integration/targets/zypper/tasks/zypper.yml (hunk 5 above)
Comment (marbud0) [reply tests/integration/targets/zypper/tasks/zypper.yml]: Done in b12916b

**Comment by marbud0** [comment]
Is there anything more I need to do for this PR to get merged?

**Comment by evrardjp** [comment]
@felixfontein could you tell us more about the next steps? I am kinda out of the loop nowadays... :/

File: plugins/modules/packaging/os/zypper.py [hunk 7]
```diff
@@ -106,6 +106,20 @@
         description:
           - Add additional options to C(zypper) command.
           - Options should be supplied in a single line as if given in the command line.
+    allow_vendor_change:
+        version_added: "2.10"
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper.py:110]: Could you remove the `version_added` here and in the next option? Right now this makes the sanity tests fail. (I have a PR where I collect all new version_added's which will be added once it is clear what the correct value is; I regularly update it so this information won't get lost.)

**Comment by felixfontein** [comment]
I'll merge this once the version_added values are gone.

File: plugins/modules/packaging/os/zypper.py (hunk 7 above)
Comment (marbud0) [reply plugins/modules/packaging/os/zypper.py]: Done

**Comment by felixfontein** [comment]
@marbud0 thanks for implementing this!
@dcermak @evrardjp thanks for reviewing!

**Pull request merged**